File: java/json-transform/src/test/java/co/nlighten/jsontransform/functions/TransformerFunctionFlatTest.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.BaseTest;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.stream.StreamSupport;

public class TransformerFunctionFlatTest extends BaseTest {
    @Test
    void testObjectFunctionFlat() {
        var arr = new String[] { "a", "b", "c"};
        var arr2 = new String[] { "d", "e", "f"};
        var arr3 = new String[][] { arr, arr2};
        var flatCombined = Arrays.stream(arr3).flatMap(x -> StreamSupport.stream(Arrays.stream(x).spliterator(), false)).toList();
        var flatArr = Arrays.stream(arr).toList();
        assertTransformation(arr3, fromJson("""
{
  "$$flat": [ "$[0]", "$[1]" ]
}
"""), flatCombined);
        assertTransformation(arr3, fromJson("""
{
  "$$flat": [ "$[0]", "$.pointingToNowhere" ]
}
"""), flatArr);
        assertTransformation(arr3, fromJson("""
{
  "$$flat": [ ["a","b","c"], ["d","e","f"] ]
}
"""), flatCombined);
        assertTransformation(arr, fromJson("""
{
  "$$flat": [ ["a","b","c"], [] ]
}
"""), flatArr);
        assertTransformation(arr, fromJson("""
{
  "$$flat": [ ["a","b","c"], null ]
}
"""), flatArr);
    }
}

File: javascript/json-transform/src/__tests__/functions/TransformerFunctionFlat.test.ts
import { describe, test } from "vitest";
import { assertTransformation } from "../BaseTransformationTest";

describe("TransformerFunctionFlat", () => {
  test("object", async () => {
    const arr = ["a", "b", "c"];
    const arr2 = ["d", "e", "f"];
    const arr3 = [arr, arr2];
    const flatCombined = arr3.flatMap(x => x);
    const flatArr = arr.slice();
    await assertTransformation(
      arr3,
      {
        $$flat: ["$[0]", "$[1]"],
      },
      flatCombined,
    );
    await assertTransformation(
      arr3,
      {
        $$flat: ["$[0]", "$.pointingToNowhere"],
      },
      flatArr,
    );
    await assertTransformation(
      arr3,
      {
        $$flat: [
          ["a", "b", "c"],
          ["d", "e", "f"],
        ],
      },
      flatCombined,
    );
    await assertTransformation(
      arr,
      {
        $$flat: [["a", "b", "c"], []],
      },
      flatArr,
    );
    await assertTransformation(
      arr,
      {
        $$flat: [["a", "b", "c"], null],
      },
      flatArr,
    );
  });
});
